libcodechecker/log/log_parser.py

```diff
@@ -0,0 +1,748 @@
+        See docs/gcc_incompatibilities.md
+        """
+
+        def contains_intrinsic_headers(include_dir):
+            """
+            Returns True if the given directory contains at least one intrinsic
+            header.
+            """
+            if not os.path.exists(include_dir):
+                return False
+            for f in os.listdir(include_dir):
+                if f.endswith("intrin.h"):
+                    return True
+            return False
+
+        result = []
+        for include_dir in include_dirs:
+            # Skip GCC's fixinclude dir
+            if os.path.basename(
+                    os.path.normpath(include_dir)) == "include-fixed":
+                continue
+            if contains_intrinsic_headers(include_dir):
+                continue
+            result.append(include_dir)
+        return result
+
+    def get_compiler_includes(self, action):
+        """
+        Returns a list of default includes of the given compiler.
+
+        action -- A BuildAction object containing the details of compilation.
+        """
+        # The first sysroot flag found among the compilation options is added
+        # to the command below to give a more precise default include list.
+        # Absence of any sysroot flags results in an empty string.
+        sysroot = ""
+        for i, item in enumerate(action.analyzer_options):
+            if item.startswith("--sysroot"):
+                # len("--sysroot") == 9
+                # len("--sysroot=") == 10
+                sysroot = action.analyzer_options[i + 1] \
+                    if len(item) == 9 else item[10:]
+                sysroot = "--sysroot=" + sysroot
+
+        # If these options are present in the original build command, they must
+        # be forwarded to get_compiler_includes and get_compiler_defines so the
+        # resulting includes point to the target that was used in the build.
+        pattern = re.compile('-m(32|64)|-std=')
+        extra_opts = filter(pattern.match, action.analyzer_options)
+        cmd = action.compiler + " " + ' '.join(extra_opts) \
+            + " -E -x " + action.lang + " " + sysroot + " - -v "
+
+        include_dirs = self.__filter_compiler_includes(
+            self.__parse_compiler_includes(self.__get_compiler_err(cmd)))
+
+        return ["-isystem " + os.path.normpath(idir) for idir in include_dirs]
+
+    def get_compiler_target(self, action):
+        """
+        Returns the target triple of the given compiler as a string.
+
+        action -- A BuildAction object containing the details of compilation.
+        """
+        lines = self.__get_compiler_err(action.compiler + ' -v')
+
+        target_label = "Target:"
+        target = ""
+
+        for line in lines.splitlines(True):
+            line = line.strip().split()
+            if len(line) > 1 and line[0] == target_label:
+                target = line[1]
+
+        return target
+
+    def get_compiler_standard(self, action):
```

Comment:
can we change this not to use the whole buildaction as an argument, just what is needed from it?